docs/architecture/adr-102-rpc-companion.md

```diff
@@ -40,109 +40,128 @@ TBD
-| `/v1/genesis`              | `/genesis`              |                                                        | Gets the genesis information              | The RPC companion service can proxy the genesis request to the full node. If there are use cases that serving the genesis from the RPC companion service (no proxy) is desirable then it can be implemented as a RPC companion service endpoint | yes   |
-| `/v1/net_info`             | `/net_info`             |                                                        | Gets network information                  | The request should proxy to the full node since the RPC companion database will not store network information                                                                                                                                   | yes   |
-| `/v1/unconfirmed_txs`      | `/unconfirmed_txs`      | * limit                                                | Gets the list of unconfirmed transactions | The request should proxy to the full node since the RPC companion database will not store unconfirmed transactions information                                                                                                                  | yes   |
-| `/v1/num_unconfirmed_txs`  | `/num_unconfirmed_txs`  |                                                        | Gets data about unconfirmed transactions  | The request should proxy to the full node since the RPC companion database will not store unconfirmed transactions information                                                                                                                  | yes   |
-| `/v1/status`               | `/status`               |                                                        | Gets node status                          | The request should proxy to the full node since the RPC companion database will not store node status information                                                                                                                               | yes   |
-
-> NOTE: The RPC Companion server instances should not implement endpoints that can modify state in the blockchain such as
-the `/broadcast_tx_*` endpoints. Since there might be many load balanced RPC server instances, this might cause issues with
-transactions, for example sequential transactions might be relayed in the wrong order causing the full node to reject some
-transactions with sequence mismatch errors. These requests will proxy to the full node as outlined above.
+| **JSONRPC method**     | *JSONRPC Parameters** | **Description**                           | **Notes**                                                                                                                                                                                                                                       | Proxy |
```

Comment:
This is outlined on line 110. The JSON-RPC interface is not something that's easy to proxy - all method calls happen via POST requests to a single endpoint.

If we don't proxy these methods, then clients need to know which endpoints to target for which requests, which complicates client architecture substantially. There is no way afaik with software like Nginx to proxy such requests, since routing would depend on data in the body of the message.

Reply:
Is it possible to switch to REST style in this API? That is, the endpoints will be selected by the query string; for "write" methods, the request method will still be POST, while read methods would be as GET.

This will enable HTTP request routing with commonly used reverse proxy/load balancer solutions.

Reply:
> Is it possible to switch to REST style in this API? That is, the endpoints will be selected by the query string; for "write" methods, the request method will still be POST, while read methods would be as GET.

We're aiming to not change the API in the first iteration of the RPC companion (accessible via the `/v1` endpoint) - we're aiming to provide as much compatibility with the existing RPC interface as possible to prevent client breakage.

We're of course free to do whatever we want with a `/v2`, but that's outside the scope of this work.

Reply:
Yes, agree with Thane, for v1 we're not changing the interfaces/methods, the client should be able to call the Companion just changing the URL.